src/sync.cpp

```diff
@@ -78,21 +78,14 @@ typedef std::map<std::pair<void*, void*>, LockStack> LockOrders;
 typedef std::set<std::pair<void*, void*> > InvLockOrders;
 
 struct LockData {
-    // Very ugly hack: as the global constructs and destructors run single
-    // threaded, we use this boolean to know whether LockData still exists,
-    // as DeleteLock can get called by global RecursiveMutex destructors
-    // after LockData disappears.
-    bool available;
-    LockData() : available(true) {}
-    ~LockData() { available = false; }
-
     LockOrders lockorders;
     InvLockOrders invlockorders;
     std::mutex dd_mutex;
 };
 LockData& GetLockData() {
-    static LockData lockdata;
-    return lockdata;
+    // See doc/developers-notes.md#construct-on-first-use-idiom
+    static auto lockdata = new LockData();
```

Comment:
```suggestion
    static LockData* lockdata = new LockData();
```

not sure why auto needs to hide both the type and the pointer-ness. The additional 5 chars don't seem overly verbose.

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/18881#issuecomment-623985636).